tests/returntypes/test_jvp.py

```diff
@@ -0,0 +1,719 @@
+
+        tape.trainable_params = {0, 1}
+        tangent = np.array([1.0, 2.0])
+
+        tapes, fn = qml.gradients.jvp(tape, tangent, param_shift)
+        assert len(tapes) == 4
+
+        res = fn(dev.batch_execute(tapes))
+        assert isinstance(res, tuple)
+
+        expected = np.array([-np.sin(x), 2 * np.cos(y)])
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_prob_expectation_values(self, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+        dev = qml.device("default.qubit", wires=2)
+        x = 0.543
+        y = -0.654
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+            qml.probs(wires=[0, 1])
+
+        tape.trainable_params = {0, 1}
+        tangent = np.array([1.0, 1.0])
+
+        tapes, fn = qml.gradients.jvp(tape, tangent, param_shift)
+        assert len(tapes) == 4
+
+        res = fn(dev.batch_execute(tapes))
+        assert isinstance(res, tuple)
+        assert len(res) == 2
+
+        expected = (
+            np.array(
+                [
+                    -2 * np.sin(x),
+                    [
+                        -(np.cos(y / 2) ** 2 * np.sin(x)) - (np.cos(x / 2) ** 2 * np.sin(y)),
+                        -(np.sin(x) * np.sin(y / 2) ** 2) + (np.cos(x / 2) ** 2 * np.sin(y)),
+                        (np.sin(x) * np.sin(y / 2) ** 2) + (np.sin(x / 2) ** 2 * np.sin(y)),
+                        (np.cos(y / 2) ** 2 * np.sin(x)) - (np.sin(x / 2) ** 2 * np.sin(y)),
+                    ],
+                ]
+            )
+            / 2
+        )
+
+        assert np.allclose(res[0], expected[0], atol=tol, rtol=0)
+        assert np.allclose(res[1], expected[1], atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("dtype", [np.float32, np.float64])
+    def test_dtype_matches_tangent(self, dtype):
+        """Tests that the vjp function matches the dtype of tangent when tangent is
+        zero-like."""
+        x = np.array([0.1], dtype=np.float64)
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(x[0], wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        dy = np.zeros(1, dtype=dtype)
+        _, func = qml.gradients.jvp(tape, dy, qml.gradients.param_shift)
+
+        assert func([]).dtype == dtype
+
+
+def expected(params):
+    x, y = 1.0 * params
+    return np.array([-np.sin(x / 2) * np.cos(x / 2), 0.0, 0.0, np.sin(x / 2) * np.cos(x / 2)])
+
+
+def ansatz(x, y):
+    qml.RX(x, wires=[0])
+    qml.RZ(y, wires=[1])
+    qml.CNOT(wires=[0, 1])
+    qml.probs(wires=[0, 1])
+
+
+class TestJVPGradients:
+    """Gradient tests for the jvp function"""
+
+    @pytest.mark.autograd
+    def test_autograd(self, tol):
+        """Tests that the output of the JVP transform
+        can be differentiated using autograd."""
```

Comment:
Is this test relevant for all interfaces? E.g., Autograd doesn't support JVPs.

Reply:
Theoretically it could, but we will not integrate it with QNode and non-backprop. It is just checking that it is diff for all int.